napari/components/layerlist.py

```diff
@@ -62,6 +62,26 @@ def selected(self):
         """List of selected layers."""
         return [layer for layer in self if layer.selected]
 
+    @property
+    def active(self):
+        """Active layer, if any, given by top most selected."""
+        active_layer = None
+        # if multiple layers are selected sets the active layer to None
+        for layer in self:
+            if active_layer is None and layer.selected:
+                active_layer = layer
+            elif active_layer is not None and layer.selected:
+                active_layer = None
+                break
+        return active_layer
```

Comment:
I see here a big performance issue. I do not like this. 
I solve a similar problem in #1945. 

There should be some cache structure (like set or dict) which will be updated base on a change of layer selection signal.

Reply:
Yeah, I was thinking about this in the context of #1945 and how we might want be able to improve it. I had started this PR before #1945 so wanted to push what I had for discussion.

> There should be some cache structure (like set or dict) which will be updated base on a change of layer selection signal.

I think this approach sounds like a good one

Reply:
Ok, I have made this use an lru_cache instead, which gets cleared when the selection status changes. This should deal with your concern. We can use a similar pattern to address the concerns of #1945 too